File: Data Structures & Algorithms/capacity-to-ship-packages-within-d-days/submission-2.java
class Solution {
    private int numdays(int[] weights, int maxw){
        int days=1;
        int curw=0;
        for(int w : weights){
            if(w+curw <=maxw){
                curw+=w;
            }else{
                days++;
                curw=w;
            }
        }
        return days;
    }
    public int shipWithinDays(int[] weights, int days) {
        int low = 0;
        int high=0;
        for(int w : weights){
            low = Math.max(low,w);
            high+=w;
        }
        int ans = high;
        while(low<=high){
            int mid = (low+high)/2;
            System.out.println(mid);
            int time = numdays(weights, mid);
            if( time<=days){
                ans=mid;
                high = mid-1;
            }
            else{
                low = mid+1;
            }
        }
        return ans;
    }
}

File: Data Structures & Algorithms/arranging-coins/submission-2.py
class Solution:
    def arrangeCoins(self, n: int) -> int:
        def solve(x):
            if x<2:
                return x
            low,high = 0, x//2
            ans = 0
            while low<=high:
                mid = (low+high)//2
                coins = mid*(mid+1)//2
                if coins==x:
                    return mid
                if coins<x:
                    ans=mid
                    low = mid+1
                else:
                    high = mid-1
            return ans
        return solve(n)